> IMO, the FIPS/non-FIPS branching for CACERTFILE is unnecessary complexity. Setting it on the connection works in both modes:
> 
> ```python
> ldap_connection = TestSyncer(self.inst.toLDAPURL())
> ldap_connection.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_DEMAND)
> ldap_connection.set_option(ldap.OPT_X_TLS_CACERTFILE, os.path.join(self.inst.get_config_dir(), "ca.crt"))
> if is_fips():
>     ldap_connection.set_option(ldap.OPT_X_TLS_PROTOCOL_MIN, ldap.OPT_X_TLS_PROTOCOL_TLS1_2)
> ldap_connection.set_option(ldap.OPT_X_TLS_NEWCTX, 0)
> ```
> 
> Otherwise, LGTM

Now I am really confused, when I try your suggestion in non FIPS mode i get:
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/home/vagrant/389-ds-base/dirsrvtests/tests/suites/syncrepl_plugin/basic_test.py", line 222, in run
    ldap_connection.set_option(ldap.OPT_X_TLS_NEWCTX, 0)
  File "/usr/lib64/python3.11/site-packages/ldap/ldapobject.py", line 981, in set_option
    return SimpleLDAPObject.set_option(self,option,invalue)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/site-packages/ldap/ldapobject.py", line 698, in set_option
    return self._ldap_call(self._l.set_option,option,invalue)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/site-packages/ldap/ldapobject.py", line 128, in _ldap_call
    result = func(*args,**kwargs)
             ^^^^^^^^^^^^^^^^^^^^
ValueError: option error
```

In non FIPS mode, I need to set the global TLS context before I rebuild it, else if fails with the above error. Per connection context rebuild works fine in FIPS mode. Is it the case that in non FIPS mode there is only a global TLS context, but in FIPS mode there is a per connection TLS context.

--- a/dirsrvtests/tests/suites/syncrepl_plugin/basic_test.py
+++ b/dirsrvtests/tests/suites/syncrepl_plugin/basic_test.py
@@ -216,17 +216,9 @@ class Sync_persist(threading.Thread, ReconnectLDAPObject, SyncreplConsumer):
         """Start a sync repl client"""
         ldap_connection = TestSyncer(self.inst.toLDAPURL())
         ldap_connection.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_DEMAND)
-
+        ldap_connection.set_option(ldap.OPT_X_TLS_CACERTFILE, os.path.join(self.inst.get_config_dir(), "ca.crt"))
         if is_fips():
-            # In FIPS mode, use connection level CA cert because a new TLS context is
-            # created, ignoring global CA cert
-            ldap_connection.set_option(ldap.OPT_X_TLS_CACERTFILE, os.path.join(self.inst.get_config_dir(), "ca.crt"))
             ldap_connection.set_option(ldap.OPT_X_TLS_PROTOCOL_MIN, ldap.OPT_X_TLS_PROTOCOL_TLS1_2)
-        else:
-            # In non FIPS mode, the connection inherits global settings, so we use global CA cert
-            ldap.set_option(ldap.OPT_X_TLS_CACERTFILE, os.path.join(self.inst.get_config_dir(), "ca.crt"))
-
-        # Rebuild TLS context
         ldap_connection.set_option(ldap.OPT_X_TLS_NEWCTX, 0)
 
         ldap_connection.simple_bind_s('cn=directory manager', 'password')
@@ -268,6 +260,7 @@ def test_sync_repl_mep(topology, request):
         5. Success
     """
     inst = topology[0]
+    inst.enable_tls()
 
     # Enable/configure retroCL
     plugin = RetroChangelogPlugin(inst)
@@ -353,6 +346,7 @@ def test_sync_repl_cookie(topology, init_sync_repl_plugins, request):
       5.: succeeds
     """
     inst = topology[0]
+    inst.enable_tls()
 
     # create a sync repl client and wait 5 seconds to be sure it is running
     sync_repl = Sync_persist(inst)
@@ -419,6 +413,8 @@ def test_sync_repl_cookie_add_del(topology, init_sync_repl_plugins, request):
       6.: succeeds
     """
     inst = topology[0]
+    inst.enable_tls()
+
     # create a sync repl client and wait 5 seconds to be sure it is running
     sync_repl = Sync_persist(inst)
     sync_repl.start()
@@ -562,6 +558,7 @@ def test_sync_repl_cenotaph(topo_m2, request):
         5. Should succeeds
     """
     m1 = topo_m2.ms["supplier1"]
+    m1.enable_tls()
     # Enable/configure retroCL
     plugin = RetroChangelogPlugin(m1)
     plugin.disable()
@@ -620,7 +617,7 @@ def test_sync_repl_dynamic_plugin(topology, request):
         3. Should succeeds
         4. Should succeeds
     """
-
+    topology.standalone.enable_tls()
     # Reset the instance in a default config
     # Disable content sync plugin
     topology.standalone.plugins.disable(name=PLUGIN_REPL_SYNC)

